File: tst/anchor.c
#include "pico.h"
#include "../check.h"

int main (void) {
    pico_init(1);
    pico_set_title("Anchoring...");

    Pico_Dim phy = {200,200};
    Pico_Dim log = { 10, 10};
    pico_set_view_raw(-1, -1, &phy, NULL, &log, NULL, NULL);

    // PIXELS
    {
        puts("centered pixel - 1dir/1baixo");
        Pico_Pos_Pct p = { 0.5, 0.50, PICO_ANCHOR_C, NULL };
        pico_output_clear();
        pico_output_draw_pixel_pct(&p);
        _pico_check("anchor-01");
    }

    {
        puts("centered pixel - 1dir/1baixo");
        Pico_Pos_Pct p = { 0.5, 0.5, PICO_ANCHOR_NW, NULL };
        pico_output_clear();
        pico_output_draw_pixel_pct(&p);
        _pico_check("anchor-02");
    }

    {
        puts("centered pixel - 1esq/1cima");
        Pico_Pos_Pct p = { 0.5, 0.5, PICO_ANCHOR_SE, NULL };
        pico_output_clear();
        pico_output_draw_pixel_pct(&p);
        _pico_check("anchor-03");
    }

    // RECTS
    {
        puts("centered rect - exact");
        Pico_Rect_Pct r = { 0.5, 0.5, 0.4, 0.4, PICO_ANCHOR_C, NULL };
        pico_output_clear();
        pico_output_draw_rect_pct(&r);
        _pico_check("anchor-04");
    }

    {
        puts("topleft centered - 1lin/1col");
        Pico_Rect_Pct r = { 0.5, 0.5, 0.4, 0.4, PICO_ANCHOR_NW, NULL };
        pico_output_clear();
        pico_output_draw_rect_pct(&r);
        _pico_check("anchor-05");
    }

    {
        puts("bottomright centered - 1lin/1col");
        Pico_Rect_Pct r = { 0.5, 0.5, 0.4, 0.4, PICO_ANCHOR_SE, NULL };
        pico_output_clear();
        pico_output_draw_rect_pct(&r);
        _pico_check("anchor-06");
    }

    {
        puts("rightmiddle centered - 1col");
        Pico_Rect_Pct r = { 0.5, 0.5, 0.4, 0.4, PICO_ANCHOR_E, NULL };
        pico_output_clear();
        pico_output_draw_rect_pct(&r);
        _pico_check("anchor-07");
    }

    {
        puts("anchor 25%25% 20%20% - 1lin/1col");
        Pico_Rect_Pct r = { 0.2, 0.2, 0.4, 0.4, {0.2,0.2}, NULL };
        pico_output_clear();
        pico_output_draw_rect_pct(&r);
        _pico_check("anchor-08");
    }

    // RECTS out of [0,100]
    {
        puts("anchor -25%-25% centered - touching border");
        Pico_Rect_Pct r = { 0.5, 0.5, 0.4, 0.4, {-0.25,-0.25}, NULL };
        pico_output_clear();
        pico_output_draw_rect_pct(&r);
        _pico_check("anchor-09");
    }

    {
        puts("anchor 125%125% centered - touching border");
        Pico_Rect_Pct r = { 0.5, 0.5, 0.4, 0.4, {1.25,1.25}, NULL };
        pico_output_clear();
        pico_output_draw_rect_pct(&r);
        _pico_check("anchor-10");
    }

    pico_init(0);
    return 0;
}

File: lua/tst/anchor.lua
local pico = require 'pico'
require 'pico.check'

pico.init(true)
pico.set.title("Anchoring...")

local phy = {w=200, h=200}
local log = {w=10, h=10}
pico.set.view(-1, phy, nil, log, nil, nil)

-- PIXELS
do
    print("centered pixel - 1dir/1baixo")
    local p = {'C', x=0.5, y=0.50}
    pico.output.clear()
    pico.output.draw.pixel(p)
    pico.check("anchor-01")
end

do
    print("centered pixel - 1dir/1baixo")
    local p = {'NW', x=0.5, y=0.5}
    pico.output.clear()
    pico.output.draw.pixel(p)
    pico.check("anchor-02")
end

do
    print("centered pixel - 1esq/1cima")
    local p = {'SE', x=0.5, y=0.5}
    pico.output.clear()
    pico.output.draw.pixel(p)
    pico.check("anchor-03")
end

-- RECTS
do
    print("centered rect - exact")
    local r = {'C', x=0.5, y=0.5, w=0.4, h=0.4}
    pico.output.clear()
    pico.output.draw.rect(r)
    pico.check("anchor-04")
end

do
    print("topleft centered - 1lin/1col")
    local r = {'NW', x=0.5, y=0.5, w=0.4, h=0.4}
    pico.output.clear()
    pico.output.draw.rect(r)
    pico.check("anchor-05")
end

do
    print("bottomright centered - 1lin/1col")
    local r = {'SE', x=0.5, y=0.5, w=0.4, h=0.4}
    pico.output.clear()
    pico.output.draw.rect(r)
    pico.check("anchor-06")
end

do
    print("rightmiddle centered - 1col")
    local r = {'E', x=0.5, y=0.5, w=0.4, h=0.4}
    pico.output.clear()
    pico.output.draw.rect(r)
    pico.check("anchor-07")
end

do
    print("anchor 25%25% 20%20% - 1lin/1col")
    local r = {{x=0.2,y=0.2}, x=0.2, y=0.2, w=0.4, h=0.4}
    pico.output.clear()
    pico.output.draw.rect(r)
    pico.check("anchor-08")
end

-- RECTS out of [0,100]
do
    print("anchor -25%-25% centered - touching border")
    local r = {{x=-0.25,y=-0.25}, x=0.5, y=0.5, w=0.4, h=0.4}
    pico.output.clear()
    pico.output.draw.rect(r)
    pico.check("anchor-09")
end

do
    print("anchor 125%125% centered - touching border")
    local r = {{x=1.25,y=1.25}, x=0.5, y=0.5, w=0.4, h=0.4}
    pico.output.clear()
    pico.output.draw.rect(r)
    pico.check("anchor-10")
end

pico.init(false)
